### ISSUE TITLE
coq: native compilation failing on macos

### ISSUE THREAD

**Issue by Alizter** [ocaml/dune#6393]
Yup, if that doesn't work then we can't merge this. Tho the error deserves its own bug:
```
+++ b/_build/.sandbox/fac578e7fc24c97bfc0799ec6c383b39/default/test/blackbox-tests/test-cases/coq/native-single.t/run.t.corrected
@@ -3,6 +3,7 @@
         coqdep foo.v.d
           coqc Nbasic_foo.{cmi,cmxs},foo.{glob,vo}
           coqc Nbasic_bar.{cmi,cmxs},bar.{glob,vo}
+  ld: warning: directory not found for option '-L$TESTCASE_ROOT/_build/default/.coq-native'
 ```
This should def not happen.

_Originally posted by @ejgallego in https://github.com/ocaml/dune/issues/6351#issuecomment-1304819764_

**Comment by Alizter** [comment]
Do we need to pass some flags to `ld` wherever that happens in Dune in order to silence this warning?

cc @rgrinberg

**Comment by rgrinberg** [comment]
I don't think so. It seems like the warning says you're passing a directory that doesn't exist to ld. Why do you do that?

**Comment by Alizter** [comment]
I suspect because we tell Dune about the `.coq-native` directory which isn't created yet. I think this was a hack from before directory targets was a thing.

**Comment by ejgallego** [comment]
The warning comes from the way Coq's `Nativelib` sets the OCaml `-I` flags when calling `ocaml` to compile the `.cmxs` files.

Indeed, that's done by the Coq's kernel, who has little idea about what the include paths should be (usually every subdir passed to Coq's `-R/-Q` option).

Usually, this include path would be managed by the build system, however at the time, Coq developers chose to do a hack `Nativelib.get_load_paths` which will actually return the list of such dirs based on what's been required; that's a horrible hack, and moreover, it doesn't allow us to distinguish which paths are coming from the current compilation context and which paths are coming from the global install setup.

Thus, for each directory in `Nativelib.get_load_paths`, Coq's will pass `-I foo` and `-I foo/.coq-native` as to cover both the install case and the non-install case.

Indeed, this mess is needed because Dune needs to set `-native-output-dir` to `.` as rules can't produce targets not in the current directory. Also, Coq used to output native files to `.coq-native` due to in-tree objects back in the day with make.
